docs/design_notes/type_promotion.md

```diff
@@ -0,0 +1,722 @@
+  x = np.arange(10, dtype=dtype)
+  assert (x + 2).dtype == dtype
+```
+
++++ {"id": "wmbZEheuVuDH"}
+
+This behavior gives motivation to our `*` notation for scalar values: the `*` is reminiscent of a wildcard that can take on any desired value.
+
+The benefit of these semantics are that you can readily express sequences of operations with clean Python code, without having to explicitly cast scalars to the appropriate type. Imagine if rather than writing this:
+```python
+3 * (x + 1) ** 2
+```
+you had to write this:
+```python
+np.int32(3) * (x + np.int32(1)) ** np.int32(2)
+```
+Although it is explicit, numerical code would become tedious to read or write. With the scalar promotion semantics described above, given an array `x` of type `int32`, the types in the second statement are implicit within the first.
+
++++ {"id": "HX7yeIf6jbjz"}
+
+## Combining Lattices
+
+Recall that we began our discussion by introducing the lattice representing type promotion within Python: `int -> float -> complex`. Let's rewrite this as `i* -> f* -> c*`, and lets further allow `i*` to subsume `u*` (after all, there is no unsigned integer scalar type in Python).
+
+Putting these all together, we get the following partial lattice representing type promotion between Python scalars and numpy arrays:
+
+```{code-cell}
+:cellView: form
+:id: koA5VFHp7tjo
+:tags: [hide-input]
+
+#@title
+import networkx as nx
+import matplotlib.pyplot as plt
+lattice = {
+  'i*': ['f*', 'u8', 'i8'], 'f*': ['c*', 'f16'], 'c*': ['c64'],
+  'u8': ['u16'], 'u16': ['u32'], 'u32': ['u64'],
+  'i8': ['i16'], 'i16': ['i32'], 'i32': ['i64'],
+  'f16': ['f32'], 'f32': ['f64'],
+  'c64': ['c128']
+}
+graph = nx.from_dict_of_lists(lattice, create_using=nx.DiGraph)
+pos = {
+  'i*': [-1.25, 0.5], 'f*': [-0.5, 2], 'c*': [0, 3],
+  'u8': [0.5, 0], 'u16': [1.5, 0], 'u32': [2.5, 0], 'u64': [3.5, 0],
+  'i8': [0, 1], 'i16': [1, 1], 'i32': [2, 1], 'i64': [3, 1],
+  'f16': [0.5, 2], 'f32': [1.5, 2], 'f64': [2.5, 2],
+  'c64': [2, 3], 'c128': [3, 3],
+}
+fig, ax = plt.subplots(figsize=(6, 5))
+nx.draw(graph, with_labels=True, node_size=1500, node_color='lightgray', pos=pos, ax=ax)
+```
+
++++ {"id": "WVjEMlmm9RS5"}
+
+Notice that this is not (yet) a true lattice: there are many pairs of nodes for which a join does not exist. However, we can think of this as a *partial* lattice, in which some pairs of nodes do not have a defined promotion behavior, and the defined portion of this partial lattice does correctly describe NumPy's array promotion behavior (leaving aside value-dependent semantics mentioned above).
+
+This sets up a nice framework by which we can think about filling-out these undefined promotion rules, by adding connections on this graph. But which connections to add?
+Broadly speaking, we want any additional connections to satisfy a few properties:
+
+1. Connections should maintain the commutativity and associativity of type promotion: that is, the graph should remain a (partial) lattice.
```

Comment:
I might phrase these as "Promotion should" rather than "Connections should".

Reply:
Done.